File: packages/election_rules/lib/src/sainte_lague.dart
import 'rule_set.dart';

/// Result of allocating [totalSeats] proportionally across lists.
class SeatAllocationResult {
  /// Seats won per list id. Every key passed in as a candidate list
  /// appears here, including lists that won zero seats.
  final Map<String, int> seatsByList;

  /// List ids excluded from allocation because their vote share fell
  /// below [ElectionRuleSet.thresholdFraction].
  final Set<String> belowThreshold;

  /// True if two or more lists tied on the deciding quotient for at least
  /// one seat during allocation. The law's tie-break procedure has not
  /// been confirmed from official text, so ties are broken deterministically
  /// (by list id) and flagged here rather than silently resolved.
  final bool tieOccurred;

  final int totalSeatsAllocated;

  const SeatAllocationResult({
    required this.seatsByList,
    required this.belowThreshold,
    required this.tieOccurred,
    required this.totalSeatsAllocated,
  });
}

/// Divisor for seat index [seatsWonSoFar] (0-based) under the standard
/// Sainte-Laguë ("odd numbers") sequence: 1, 3, 5, 7, ...
double _standardDivisor(int seatsWonSoFar) => (2 * seatsWonSoFar + 1).toDouble();

/// Divisor for the "modified" Sainte-Laguë sequence used by some countries
/// to make it harder for very small lists to win a first seat: 1.4, 3, 5, 7, ...
double _modifiedDivisor(int seatsWonSoFar) => seatsWonSoFar == 0 ? 1.4 : _standardDivisor(seatsWonSoFar);

/// Allocates [rules.generalPoolSeats] seats across [votesByList] using the
/// highest-averages Sainte-Laguë method specified by [rules].
///
/// This function is deliberately pure and deterministic: given the same
/// votes and rule set it always returns the same result, which is required
/// for forecast reproducibility (a Monte Carlo simulation calls this many
/// thousands of times per run).
///
/// Lists whose vote share is strictly below [ElectionRuleSet.thresholdFraction]
/// are excluded from the allocation pool entirely (they receive zero seats)
/// but their votes still count toward total valid votes when computing
/// vote shares, matching how PR thresholds normally operate.
SeatAllocationResult allocateSeatsSainteLague({
  required Map<String, int> votesByList,
  required ElectionRuleSet rules,
}) {
  if (rules.allocationMethod != AllocationMethod.sainteLague &&
      rules.allocationMethod != AllocationMethod.modifiedSainteLague) {
    throw UnsupportedError(
      'allocateSeatsSainteLague only supports sainteLague/modifiedSainteLague; '
      'rule set ${rules.id} specifies ${rules.allocationMethod}.',
    );
  }

  final divisorFn =
      rules.allocationMethod == AllocationMethod.modifiedSainteLague ? _modifiedDivisor : _standardDivisor;

  final totalVotes = votesByList.values.fold<int>(0, (a, b) => a + b);
  final seats = <String, int>{for (final id in votesByList.keys) id: 0};
  final belowThreshold = <String>{};

  if (totalVotes <= 0) {
    return SeatAllocationResult(
      seatsByList: seats,
      belowThreshold: votesByList.keys.toSet(),
      tieOccurred: false,
      totalSeatsAllocated: 0,
    );
  }

  final eligible = <String>[];
  for (final entry in votesByList.entries) {
    final share = entry.value / totalVotes;
    if (share < rules.thresholdFraction) {
      belowThreshold.add(entry.key);
    } else {
      eligible.add(entry.key);
    }
  }

  final seatsToAllocate = rules.generalPoolSeats;
  bool tieOccurred = false;

  for (var seatIndex = 0; seatIndex < seatsToAllocate; seatIndex++) {
    if (eligible.isEmpty) break;

    String? winner;
    double bestQuotient = -1;
    var tiedThisRound = false;

    // Deterministic order (sorted list ids) so ties break the same way
    // every run, which matters for reproducibility across simulations.
    final ordered = [...eligible]..sort();
    for (final listId in ordered) {
      final quotient = votesByList[listId]! / divisorFn(seats[listId]!);
      if (quotient > bestQuotient) {
        bestQuotient = quotient;
        winner = listId;
        tiedThisRound = false;
      } else if (quotient == bestQuotient) {
        tiedThisRound = true;
      }
    }

    if (tiedThisRound) tieOccurred = true;
    seats[winner!] = seats[winner]! + 1;
  }

  final allocated = seats.values.fold<int>(0, (a, b) => a + b);

  return SeatAllocationResult(
    seatsByList: seats,
    belowThreshold: belowThreshold,
    tieOccurred: tieOccurred,
    totalSeatsAllocated: allocated,
  );
}

File: packages/election_rules_py/election_rules_py/sainte_lague.py
"""Sainte-Lague highest-averages seat allocation.

Direct port of packages/election_rules/lib/src/sainte_lague.dart — keep in
sync; that Dart version is the one covered by `dart test` in this repo's
sandbox.
"""

from __future__ import annotations

from dataclasses import dataclass

from .rule_set import AllocationMethod, ElectionRuleSet


@dataclass(frozen=True)
class SeatAllocationResult:
    seats_by_list: dict[str, int]
    below_threshold: set[str]
    tie_occurred: bool
    total_seats_allocated: int


def _standard_divisor(seats_won_so_far: int) -> float:
    return float(2 * seats_won_so_far + 1)


def _modified_divisor(seats_won_so_far: int) -> float:
    return 1.4 if seats_won_so_far == 0 else _standard_divisor(seats_won_so_far)


def allocate_seats_sainte_lague(
    votes_by_list: dict[str, int],
    rules: ElectionRuleSet,
) -> SeatAllocationResult:
    """Allocate rules.general_pool_seats seats across votes_by_list.

    Pure and deterministic: same inputs always produce the same result,
    which forecast reproducibility depends on (a Monte Carlo run calls
    this tens of thousands of times).

    Lists whose vote share is strictly below rules.threshold_fraction are
    excluded from allocation (zero seats) but still count toward total
    valid votes when computing shares.
    """
    if rules.allocation_method not in (
        AllocationMethod.SAINTE_LAGUE,
        AllocationMethod.MODIFIED_SAINTE_LAGUE,
    ):
        raise NotImplementedError(
            f"allocate_seats_sainte_lague only supports sainte_lague/"
            f"modified_sainte_lague; rule set {rules.id} specifies "
            f"{rules.allocation_method}."
        )

    divisor_fn = (
        _modified_divisor
        if rules.allocation_method == AllocationMethod.MODIFIED_SAINTE_LAGUE
        else _standard_divisor
    )

    total_votes = sum(votes_by_list.values())
    seats: dict[str, int] = {list_id: 0 for list_id in votes_by_list}
    below_threshold: set[str] = set()

    if total_votes <= 0:
        return SeatAllocationResult(
            seats_by_list=seats,
            below_threshold=set(votes_by_list.keys()),
            tie_occurred=False,
            total_seats_allocated=0,
        )

    eligible: list[str] = []
    for list_id, votes in votes_by_list.items():
        share = votes / total_votes
        if share < rules.threshold_fraction:
            below_threshold.add(list_id)
        else:
            eligible.append(list_id)

    seats_to_allocate = rules.general_pool_seats
    tie_occurred = False

    for _ in range(seats_to_allocate):
        if not eligible:
            break

        winner: str | None = None
        best_quotient = -1.0
        tied_this_round = False

        for list_id in sorted(eligible):
            quotient = votes_by_list[list_id] / divisor_fn(seats[list_id])
            if quotient > best_quotient:
                best_quotient = quotient
                winner = list_id
                tied_this_round = False
            elif quotient == best_quotient:
                tied_this_round = True

        if tied_this_round:
            tie_occurred = True
        assert winner is not None
        seats[winner] += 1

    allocated = sum(seats.values())

    return SeatAllocationResult(
        seats_by_list=seats,
        below_threshold=below_threshold,
        tie_occurred=tie_occurred,
        total_seats_allocated=allocated,
    )
